proposals/0000-pure-template-haskell.md

```diff
@@ -0,0 +1,294 @@
+Users can use `addDependentFile` to declare a dependency on an external file, giving a limited improvement to recompilation checking.
+
+Pure Template Haskell would help in that the current recompilation check would actually be correct for modules that only have Pure Template Haskell. 
+We would still have the same difficulty for modules with impure Template Haskell, but since there would be far fewer, we could potentially use a more conservative recompilation check without having a terrible impact on the innocent majority.
+
+### Better motivation
+
+The above motivation is somewhat weak. 
+Alone, this proposal is more of a “cleanliness” proposal. However, the author believes it is an important step towards making TH easier for GHC to work with, which in the long run will be quite beneficial. 
+The motivation for this proposal is therefore missing some support from an (unwritten) larger manifesto on this topic.
+
+## Proposed Change Specification
+
+### 0. Terminology
+
+A Template Haskell quote is said to be "impure" if it has type `m a` where `m` has or requires a `MonadIO` instance.
+
+A Template Haskell splice is said to be "impure" if the code inside the splice has type `m a` where `m` has or requires a `MonadIO` instance.
+
+Note that today types for quotes and splice code are already inferred with appropriate constraints depending on what they use.
+
+### 1. Syntax
+
+A new syntax for top-level “impure splices” is introduced.
+
+- Impure untyped splice: `$!..`
+- Impure typed splice: `$$!..`
+
+We will elide the distinction between a splice being impure in the sense of "needing IO" and a splice being impure in the sense of "using impure splice syntax", so long as this is not confusing.
+
+Splices which do not use the impure splice syntax are called “normal” splices. This includes:
+- Splices of all kinds when `EnforceTemplateHaskellPurity` is disabled
+- Pure splices when `EnforceTemplateHaskellPurity` is enabled
+- Nested (non-top-level) splices
+
+### 2. Additions to `template-haskell`
+
+Introduce a new class `Reify`, containing the `reify` methods from `Quasi`:
+```
+–- could use MonadError here, but we’re going to impose a Reify m constraint at 
+-- the top level, and we probably want users to be able to pick their own 
+-- error monad for m if necessary
+class Monad m => Reify m where
+  reify          :: Name -> ExceptT ReifyError m Info
+  reifyFixity    :: Name -> ExceptT ReifyError m (Maybe Fixity)
+  reifyType      :: Name -> ExceptT ReifyError m Type
+  reifyInstances :: Name -> [Type] -> ExceptT ReifyError m [Dec]
+  reifyRoles         :: Name -> ExceptT ReifyError m [Role]
+  reifyAnnotations   :: Data a => AnnLookup -> ExceptT ReifyError m [a]
+  reifyModule        :: Module -> ExceptT ReifyError m ModuleInfo
+  reifyConStrictness :: Name -> ExceptT ReifyError m [DecidedStrictness]
+```
+
+Introduce a type `PureQ`, which parallels `Q` but only has access to `Quote` and `Reify`.
```

Comment:
I think there will be an explicit coercion function to turn `PureQ` into `Q`, right? How else would you call pure splices from an impure one? 

What if I want to call a pure splice function `p :: (PureQ a -> PureQ a) -> PureQ a` in impure `Q a` that accepts a splice callback? Today, I can easily call such a function; not so with `PureQ`.